File: schedule_forge/scheduling/generator.py
"""Generation of the conflict free exam systems (requirement 1.2).

The generator does not search over all the exams at once. It first splits them
into independent components (see `decomposition`), and then walks the Cartesian
product of the component solutions like an odometer: the last component turns on
every step, the one before it only when the last one has gone round.

Two properties follow from that, and they are the whole point of the engine:

* the exact number of exam systems is the product of the component counts. It
  is known before a single system is built, so the software can state how large
  the set of possibilities really is instead of guessing;
* producing one more exam system costs one step of the last component plus the
  copy of the placement - it does not cost a walk over all the exams.

The number of exam systems still grows exponentially with the number of exams
(D free dates and N exams allow up to D**N systems), so the systems are produced
lazily and the caller states how many it wants and how long it may take;
requirement 5.1 gives the default budget.
"""

import time

from ..model.exam import ExamSystem, ScheduledExam
from .constraints import NoTwoExamsSameDayInYearAndProgram
from .decomposition import ProblemDecomposition


class GenerationReport(object):
    """What the generator did: how many systems, out of how many, and why it stopped."""

    COMPLETE = "complete"
    REACHED_LIMIT = "reached limit"
    TIMED_OUT = "timed out"

    __slots__ = ("produced", "status", "seconds", "total")

    def __init__(self, total=None):
        self.produced = 0
        self.status = self.COMPLETE
        self.seconds = 0.0
        self.total = total

    @property
    def is_complete(self):
        return self.status == self.COMPLETE

    @property
    def total_text(self):
        if self.total is None:
            return "an unknown number of"
        return "{:,}".format(self.total)

    def describe(self):
        if self.status == self.COMPLETE:
            return ("all %d possible exam systems were generated (%.2f seconds)"
                    % (self.produced, self.seconds))
        if self.status == self.REACHED_LIMIT:
            return ("the first %d of %s possible exam systems were generated "
                    "(%.2f seconds)"
                    % (self.produced, self.total_text, self.seconds))
        return ("%d of %s possible exam systems were generated before the time "
                "limit of the run was reached (%.2f seconds)"
                % (self.produced, self.total_text, self.seconds))


class ExamSystemGenerator(object):
    """Produces the conflict free exam systems of a list of exams."""

    #: How many systems are produced between two readings of the clock.
    _TIME_CHECK_INTERVAL = 512

    def __init__(self, exams, periods, constraints=None, availability=None,
                 pruner=None, diversify=False):
        self.exams = list(exams)
        self.periods = periods
        self.constraints = (list(constraints) if constraints is not None
                            else [NoTwoExamsSameDayInYearAndProgram()])
        self.availability = availability
        # When the user asks for the best exam systems, the components and
        # their dates are shuffled, so a search bounded by a budget samples a
        # different slice of the space on every run instead of always the
        # same corner of it (see `ProblemDecomposition`). Plain listing
        # (versions 1.0 and 2.0) leaves it off, which is also what keeps the
        # guarantee that the very first system places every exam as early as
        # it can.
        self.decomposition = ProblemDecomposition(self.exams, self.periods,
                                                  self.constraints, availability,
                                                  shuffle=diversify)
        #: Optional: throws a half built system away as soon as a count over the
        #: whole system is broken (version 3.0, see `scheduling.partial`).
        self.pruner = pruner
        self.report = GenerationReport(self.total_systems())
        self._scheduled = {}

    def total_systems(self):
        """The exact number of possible exam systems, or None when unknown.

        Costs no search: it is the product of the counts of the components.
        """
        return self.decomposition.total_systems()

    def components(self):
        return self.decomposition.components

    def candidate_dates(self, exam):
        """The dates that exam may take, earliest one first.

        The dates of its exam period, without the ones its instructor is not
        available on.
        """
        return self.decomposition.dates_of_exam[self.exams.index(exam)]

    def generate(self, max_systems=None, time_limit_seconds=None):
        """Yield `ExamSystem` objects until they run out, or a budget is spent.

        `max_systems` limits how many systems are produced, `time_limit_seconds`
        how long the generation may run. Both may be None, meaning no limit.
        """
        started = time.time()
        report = GenerationReport(self.total_systems())
        self.report = report
        deadline = (started + time_limit_seconds
                    if time_limit_seconds is not None else None)

        try:
            for system in self._walk(report, max_systems, deadline):
                yield system
        finally:
            report.seconds = time.time() - started

    def depth_of_position(self):
        """Which component every exam belongs to, by its index in the exam list."""
        depths = [0] * len(self.exams)
        for depth, component in enumerate(self.decomposition.components):
            for position in component.positions:
                depths[position] = depth
        return depths

    def _walk(self, report, max_systems, deadline):
        components = self.decomposition.components
        depth_count = len(components)
        if not self.exams or depth_count == 0:
            return
        if self.total_systems() == 0:
            # One component alone has no solution, so no exam system exists.
            # Finding that out here costs nothing; a search over all the exams
            # would rediscover it once for every placement of the other exams.
            return

        pruner = self.pruner
        if pruner is not None:
            pruner.reset()
        placement = [None] * len(self.exams)
        scheduled_of = self._scheduled_exam
        iterators = [None] * depth_count
        iterators[0] = components[0].iter_solutions()
        last = depth_count - 1
        produced = 0
        examined = 0
        depth = 0
        applied_to = -1

        while depth >= 0:
            solution = next(iterators[depth], None)
            if solution is None:
                depth -= 1
                continue

            if pruner is not None:
                while applied_to >= depth:
                    pruner.unapply(applied_to)
                    applied_to -= 1

            component = components[depth]
            positions = component.positions
            dates = component.dates
            pairs = [(positions[place], dates[solution[place]])
                     for place in range(len(positions))]
            for position, date in pairs:
                placement[position] = scheduled_of(position, date)

            if pruner is not None:
                examined += 1
                if not pruner.apply(depth, pairs):
                    # Every system below this placement breaks the same count.
                    pruner.unapply(depth)
                    if deadline is not None and not examined % self._TIME_CHECK_INTERVAL:
                        if time.time() >= deadline:
                            report.status = GenerationReport.TIMED_OUT
                            return
                    continue
                applied_to = depth

            if depth != last:
                depth += 1
                iterators[depth] = components[depth].iter_solutions()
                continue

            produced += 1
            report.produced = produced
            yield ExamSystem(list(placement))

            if max_systems is not None and produced >= max_systems:
                report.status = GenerationReport.REACHED_LIMIT
                return
            if deadline is not None and not produced % self._TIME_CHECK_INTERVAL:
                if time.time() >= deadline:
                    report.status = GenerationReport.TIMED_OUT
                    return

    def _scheduled_exam(self, position, date):
        """A `ScheduledExam` per exam and date, built once and then reused.

        `ScheduledExam` is a value that is never changed, so the same object can
        appear in many exam systems. It keeps the cost of one more system at the
        copy of the placement list.
        """
        key = (position, date)
        scheduled = self._scheduled.get(key)
        if scheduled is None:
            scheduled = ScheduledExam(self.exams[position], date)
            self._scheduled[key] = scheduled
        return scheduled

File: web/src/engine/generator.ts
/**
 * Finding the exam systems the user asked for (version 3.0, sections 2 and 3).
 *
 * The run has three steps, and they are deliberately separate:
 *
 * 1. the walk produces exam systems. The thresholds that relate two exams -
 *    2.1 and 2.2 - are already inside the decomposition, so a system that
 *    breaks them is never built, and the counts over a whole system - 2.3, 2.4,
 *    2.5 and the room capacity - drop a placement while it is still partial;
 * 2. every system that comes out is measured, and, when at least one sorting
 *    criterion is on, weighed against the worst of the best `maxCandidates`
 *    found so far (`BoundedBest`) - so the search never stops merely because it
 *    filled its quota, only because it ran out of systems to look at, examined
 *    as many as `maxExamined` allows, or ran out of time. What is kept is
 *    therefore the best `maxCandidates` systems out of everything the search
 *    looked at, not the first ones that happened to pass;
 * 3. the kept systems are put in order by the criteria of section 3. That last
 *    step is cheap and is done again whenever the user only changes the order
 *    of the criteria, which is exactly what the requirement expects: the
 *    sorting may change during a run, the thresholds may not.
 *
 * When no criterion is on, "best" has no meaning to search for, and the run
 * falls back to keeping the first `maxCandidates` systems that pass the
 * thresholds - there is nothing to prefer one over another for.
 */

import { Component, Decomposition, iterSolutions } from "./decomposition";
import { Exam, ExamSystem, FacultyRules, Room, ScheduledExam } from "./model";
import { PartialThresholdChecker } from "./partial";
import { RoomAllocation, RoomAllocator } from "./rooms";
import { BoundedBest } from "./topk";
import { SystemMetrics, compareByCriteria, measure, passesThresholds } from "./quality";
import { Settings, SortCriterion } from "./settings";
import { translate as t } from "../i18n/translate";

export interface Candidate {
  system: ExamSystem;
  metrics: SystemMetrics;
  allocation: RoomAllocation | null;
}

export type SearchStatus = "complete" | "enough" | "examined limit" | "timed out";

export interface SearchReport {
  examined: number;
  accepted: number;
  status: SearchStatus;
  seconds: number;
  /** The number of systems before the threshold requirements, or null. */
  totalSystems: bigint | null;
}

export interface SearchResult {
  candidates: Candidate[];
  report: SearchReport;
}

export function describeSearch(report: SearchReport, kept: number): string {
  const total =
    report.totalSystems === null ? t("search.unknownCount") : report.totalSystems.toLocaleString("en-US");
  const examined = report.examined.toLocaleString("en-US");
  let head: string;
  if (report.status === "complete") {
    head = t("search.headComplete", { total });
  } else if (report.status === "enough") {
    head = t("search.headEnough", { examined, total });
  } else if (report.status === "examined limit") {
    head = t("search.headExaminedLimit", { examined, total });
  } else {
    head = t("search.headTimedOut", { examined, total });
  }
  const tail =
    kept < report.accepted
      ? t("search.tailWithKept", { accepted: report.accepted, kept })
      : t("search.tailAllKept", { accepted: report.accepted });
  return t("search.summary", { head, tail, seconds: report.seconds.toFixed(2) });
}

export interface SearchInput {
  exams: Exam[];
  decomposition: Decomposition;
  settings: Settings;
  rooms?: Room[];
  faculty?: FacultyRules;
}

export function runSearch(input: SearchInput): SearchResult {
  const started = performance.now();
  const { exams, decomposition, settings } = input;
  const components = decomposition.components;

  const allocator = input.rooms && input.rooms.length
    ? new RoomAllocator(input.rooms, settings.defaultStudents)
    : null;
  const pruner = new PartialThresholdChecker(
    exams,
    decomposition.depthOfPosition,
    settings,
    allocator ? allocator.totalCapacity : null
  );
  const usePruner = pruner.isNeeded;
  if (usePruner) pruner.reset();

  const report: SearchReport = {
    examined: 0,
    accepted: 0,
    status: "complete",
    seconds: 0,
    totalSystems: decomposition.total,
  };

  if (!exams.length || !components.length || decomposition.total === 0n) {
    report.seconds = (performance.now() - started) / 1000;
    return { candidates: [], report };
  }

  const deadline = settings.timeLimitSeconds
    ? started + settings.timeLimitSeconds * 1000
    : null;
  // When the user asks for the best exam systems, `decompose` (the caller's
  // job, before this function is ever called) has already shuffled the
  // components and their dates, so the standard walk below - unchanged -
  // samples a different neighbourhood of the search space on every run
  // instead of always exhausting the last component before trying another
  // value of the first.
  const ranked = settings.sortCriteria.length > 0;
  const best = ranked
    ? new BoundedBest<Candidate>(settings.maxCandidates, (a, b) =>
        compareByCriteria(a.metrics, b.metrics, settings.sortCriteria) < 0
      )
    : null;
  const firstFound: Candidate[] = [];
  let accepted = 0;
  const seen = new Set<string>();
  const placement = new Array<string>(exams.length).fill("");
  const iterators: Generator<number[]>[] = new Array(components.length);
  const last = components.length - 1;
  let depth = 0;
  let appliedTo = -1;
  let examined = 0;
  iterators[0] = iterSolutions(components[0]);

  walk: while (depth >= 0) {
    const step = iterators[depth].next();
    if (step.done) {
      depth -= 1;
      continue;
    }

    if (usePruner) {
      while (appliedTo >= depth) {
        pruner.unapply(appliedTo);
        appliedTo -= 1;
      }
    }

    const component: Component = components[depth];
    const pairs: [number, string][] = component.positions.map((position, place) => [
      position,
      component.dates[step.value[place]],
    ]);
    for (const [position, date] of pairs) placement[position] = date;

    if (usePruner) {
      if (!pruner.apply(depth, pairs)) {
        // Every system below this placement breaks the same count.
        pruner.unapply(depth);
        continue;
      }
      appliedTo = depth;
    }

    if (depth !== last) {
      depth += 1;
      iterators[depth] = iterSolutions(components[depth]);
      continue;
    }

    examined += 1;
    const key = placement.join("|");
    if (!seen.has(key)) {
      // A spread out walk may reach one system by two ways, and the user must
      // never be shown the same system twice.
      seen.add(key);
      const system: ExamSystem = exams.map<ScheduledExam>((exam, position) => ({
        exam,
        date: placement[position],
      }));
      const metrics = measure(system);
      if (passesThresholds(metrics, settings)) {
        const allocation = allocator ? allocator.allocate(system) : null;
        if (!(settings.requireRooms && allocation && !allocation.isComplete)) {
          accepted += 1;
          const candidate: Candidate = { system, metrics, allocation };
          if (best) {
            best.offer(candidate);
          } else {
            firstFound.push(candidate);
            if (settings.maxCandidates && firstFound.length >= settings.maxCandidates) {
              report.status = "enough";
              break walk;
            }
          }
        }
      }
    }

    if (settings.maxExamined && examined >= settings.maxExamined) {
      report.status = "examined limit";
      break walk;
    }
    if (deadline !== null && examined % 256 === 0 && performance.now() >= deadline) {
      report.status = "timed out";
      break walk;
    }
  }

  const candidates = best ? [...best.items] : firstFound;
  report.examined = examined;
  report.accepted = accepted;
  sortCandidates(candidates, settings.sortCriteria);
  report.seconds = (performance.now() - started) / 1000;
  return { candidates, report };
}

/** Order the candidates again, by the criteria of section 3. */
export function sortCandidates(candidates: Candidate[], criteria: SortCriterion[]): Candidate[] {
  if (criteria.length) {
    candidates.sort((first, second) =>
      compareByCriteria(first.metrics, second.metrics, criteria)
    );
  }
  return candidates;
}
